docs/tutorials/data_sources/pytorch_dataset_tutorial.md

```diff
@@ -0,0 +1,164 @@
+---
+
+# PyTorch dataset tutorial
+
++++
+
+In this tutorial we're going to learn how to migrate from Torchvision to Grain pipeline for [Fashion-MNIST](https://github.com/zalandoresearch/fashion-mnist) dataset.
+
++++
+
+## Setup
+
+Here we require `grain`, `torch`, and `torchvision` dependencies installed but the last will be used primarily for dataset access. Additionally we want `matplotlib` in our environment for displaying samples.
+
+```{code-cell} ipython3
+# @test {"output": "ignore"}
+!pip install grain
+# @test {"output": "ignore"}
+!pip install numpy matplotlib torch torchvision
+```
+
+```{code-cell} ipython3
+import grain
+import numpy as np
+import matplotlib.pyplot as plt
+import torch
+from PIL.Image import Image
+
+# PyTorch imports
+from torchvision import datasets
+from torchvision.transforms import ToTensor, Lambda
+
+rng = np.random.default_rng(0)
+```
+
+## Loading dataset
+
+`FashionMNIST` function provides access to our dataset. There are 60k samples in the dataset, where each one of them is a Pillow image instance together with a label.
+
+```{code-cell} ipython3
+fashion_mnist = datasets.FashionMNIST(root="data", train=True, download=True)
+print(fashion_mnist)
+print(fashion_mnist[0])
+fashion_mnist[0][0]
+```
+
+Here, for example, we have an ankle boot (label 9). First we should acknowledge that both sample and label requires a dedicated preprocessing:
+- Sample should be a PyTorch tensor
+- Label needs to use one-hot encoding, so `9` becomes `[0,0,0,0,0,0,0,0,0,1]`
+
+Let's inspect a few more samples first.
+
+```{code-cell} ipython3
+nrows, ncols = 4, 8
+fig, axs = plt.subplots(nrows, ncols)
+for x in range(ncols):
+    for y in range(nrows):
+        axs[y, x].imshow(fashion_mnist.data[rng.integers(len(fashion_mnist))])
+        axs[y, x].set_axis_off()
+```
+
+`torchvision`'s function, which imports the dataset, already offers `transform` and `target_transform` arguments. Let's confirm that `ToTensor` and `Lambda` do what we anticipate.
+
+```{code-cell} ipython3
+fashion_mnist_2 = datasets.FashionMNIST(
+    root="data",
+    train=True,
+    download=True,
+    transform=ToTensor(),
+    target_transform=Lambda(
+        lambda y: torch.zeros(10, dtype=torch.float).scatter_(0, torch.tensor(y), value=1)
+    )
+)
+```
+
+```{code-cell} ipython3
+(fashion_mnist_2[0][0].size(), fashion_mnist_2[0][1])
+```
+
+The resulting contents are PyTorch tensor and one-hot encoded label indeed.
+
+Now let's move to Grain! Given the unprocessed dataset we want to arrive with the same transformation. You might have noticed in the "Dataset basics" tutorial that defining a data source requires implementing a class that inherits from `grain.sources.RandomAccessDataSource`. With PyTorch datasets we don't need it - the dataset is already compliant with the protocol.
+
+First let's define our data source with random access capabilities.
+
+```{code-cell} ipython3
+fashion_source = fashion_mnist
+```
+
+## Preprocessing
```

Comment:
If the previous suggestion is accepted, maybe this is not needed?

Reply:
That's correct - done!